pkg/shell-operator/bootstrap.go

```diff
@@ -0,0 +1,237 @@
+	if err != nil {
+		return fmt.Errorf("start HTTP server for hook metrics: %s", err)
+	}
+	if hookMetricStorage == nil {
+		op.HookMetricStorage = op.MetricStorage
+	}
+
+	// 'main' Kubernetes client.
+	op.KubeClient, err = InitDefaultMainKubeClient(metricStorage)
+	if err != nil {
+		return err
+	}
+
+	// ObjectPatcher with a separate Kubernetes client.
+	op.ObjectPatcher, err = InitDefaultObjectPatcher(metricStorage)
+	if err != nil {
+		return err
+	}
+
+	SetupEventManagers(op)
+
+	return nil
+}
+
+// AssembleShellOperator uses settings in app package to create all
+// dependencies needed for the full-fledged ShellOperator.
+//
+// - check directories
+// - start debug server
+// - initialize dependencies:
+//   - metric storage
+//   - kubernetes client config
+//   - empty set of task queues
+//   - hook manager
+//   - kubernetes events manager
+//   - schedule manager
+func AssembleShellOperator(op *ShellOperator, hooksDir string, tempDir string, debugServer *debug.Server, runtimeConfig *config.Config) (err error) {
+	RegisterDefaultRoutes(op)
+
+	RegisterDebugQueueRoutes(debugServer, op)
+	RegisterDebugHookRoutes(debugServer, op)
+	RegisterDebugConfigRoutes(debugServer, runtimeConfig)
+
+	RegisterShellOperatorMetrics(op.MetricStorage)
+
+	// Create webhookManagers with dependencies.
+	SetupHookManagers(op, hooksDir, tempDir)
+
+	// Search and configure all hooks.
+	err = op.InitHookManager()
+	if err != nil {
+		return fmt.Errorf("initialize HookManager fail: %s", err)
+	}
+
+	// Load validation hooks.
+	err = op.InitValidatingWebhookManager()
+	if err != nil {
+		return fmt.Errorf("initialize ValidatingWebhookManager fail: %s", err)
+	}
+
+	// Load conversion hooks.
+	err = op.InitConversionWebhookManager()
+	if err != nil {
+		return fmt.Errorf("initialize ConversionWebhookManager fail: %s", err)
+	}
+
+	return nil
+}
+
+// SetupEventManagers instantiate queues and managers for schedule and Kubernetes events.
+func SetupEventManagers(op *ShellOperator) {
+	// Initialize the task queues set with the "main" queue.
+	op.TaskQueues = queue.NewTaskQueueSet()
+	op.TaskQueues.WithContext(op.ctx)
+	op.TaskQueues.WithMetricStorage(op.MetricStorage)
+
+	// Initialize schedule manager.
+	op.ScheduleManager = schedule_manager.NewScheduleManager()
+	op.ScheduleManager.WithContext(op.ctx)
+
+	// Initialize kubernetes events manager.
+	op.KubeEventsManager = kube_events_manager.NewKubeEventsManager()
+	op.KubeEventsManager.WithKubeClient(op.KubeClient)
+	op.KubeEventsManager.WithContext(op.ctx)
+	op.KubeEventsManager.WithMetricStorage(op.MetricStorage)
+
+	// Initialize events handler that emit tasks to run hooks
+	op.ManagerEventsHandler = NewManagerEventsHandler()
+	op.ManagerEventsHandler.WithContext(op.ctx)
+	op.ManagerEventsHandler.WithTaskQueueSet(op.TaskQueues)
+	op.ManagerEventsHandler.WithScheduleManager(op.ScheduleManager)
+	op.ManagerEventsHandler.WithKubeEventsManager(op.KubeEventsManager)
```

Comment:
Most of these `WithXXX` looks weird honestly :( They are required for workability but you can miss this call. Don't think we should do it at the moment but it could be a point in future. To make it as required input args for constructor

Reply:
Yes, one day we'll refactor these "conventions" as well. This PR only takes out initialization responsibility from ShellOperator's methods.